onnx/defs/nn/defs.cc

```diff
@@ -2720,22 +2720,35 @@ where the mean and variance are computed per instance per group of channels, and
 groups `num_groups` should be divisible by the number of channels so that there are
 an equal number of channels per group.
 
+The overall computation has two stages: the first stage normalizes the elements to
+have zero mean and unit variance for each instance in each group, and the second
+stage scales and shifts the results of the first stage. The floating-point precision
+used in the first stage is determined by the `stash_type` attribute. For example,
+if `stash_type` is 1, the operator casts all input variables to 32-bit float,
+performs the computation, and finally casts the normalized results back to the
+original type of `X`. The second stage does not edepend on `stash_type`.
+
 When the number of groups is the same as the number of channels, this operator is
 equivalent to InstanceNormalization. When there is only one group, this operator
 is equivalent to LayerNormalization.
 )DOC";
 
 ONNX_OPERATOR_SET_SCHEMA(
     GroupNormalization,
-    18,
+    21,
     OpSchema()
-        .SetDoc(GroupNormalization_ver18_doc)
+        .SetDoc(GroupNormalization_ver21_doc)
         .Attr("epsilon", "The epsilon value to use to avoid division by zero.", AttributeProto::FLOAT, 1e-5f)
         .Attr(
             "num_groups",
             "The number of groups of channels. It should be a divisor of the number of channels `C`.",
             AttributeProto::INT,
             true)
+        .Attr(
+            "stash_type",
+            "The floating-point precision used in stage one of the computation.",
```

Comment:
May be clarify the allowed types (bfloat1 and float) here and maybe in documentation above also.

Reply:
On second thought, I think we should relax the restriction on stash_type to allow any dtype that the input can take. The motivation for having a stash_type is so that if necessary the user can choose to use a higher precision than input dtype for stage 1 computations. But I don't think the spec itself should prohibit user from using any desired floating point type.

In the LayerNorm case, the spec also similarly doesn't put any restrictions, but the function definition itself asserts stash_type to be float or bfloat16, which IMO is overly restrictive.

Reply:
The proposed generalization seems ok to me. Is that float/blfoat16/float16/double ? I suppose we actually would need this for backward compatibility (when doing version-conversion from older version)?

Reply:
@gramalingam Yes, it would be the same set of dtypes as input, so float/blfoat16/float16/double.
And yes, backward compatibility is a good point too.